WordPress/openverse-api
# Reduce DB queries needed in search results

## Fixes


Fixes #1036 by @sarayourfriend 

## Description


Currently a search result needs several DB queries to populate the page.
1. Get `Hit` instances from ES
2. Get DB rows corresponding to the hits.
3. For each audio row, search for `AudioSet` and `MatureAudio` instances. For each image search for the `MatureImage` instances.

This PR
- uses ORM-fu to bring the total queries down to 1!
- improves the structure of reporting models to use FK 
- improves the admin experience using autocomplete field
- uses `OpenledgerModel` in more places for consistent timestamping fields

This is an example query for audio, it includes the related `matureaudio` and `audioset` data too!

```sql
SELECT 
  "audio"."id", 
  "audio"."created_on", 
  "audio"."updated_on", 
  "audio"."identifier", 
  "audio"."foreign_identifier", 
  "audio"."title", 
  "audio"."foreign_landing_url", 
  "audio"."creator", 
  "audio"."creator_url", 
  "audio"."thumbnail", 
  "audio"."provider", 
  "audio"."url", 
  "audio"."filesize", 
  "audio"."filetype", 
  "audio"."watermarked", 
  "audio"."license", 
  "audio"."license_version", 
  "audio"."source", 
  "audio"."last_synced_with_source", 
  "audio"."removed_from_source", 
  "audio"."view_count", 
  "audio"."tags", 
  "audio"."tags_list", 
  "audio"."category", 
  "audio"."meta_data", 
  "audio"."bit_rate", 
  "audio"."sample_rate", 
  "audio"."audio_set_foreign_identifier", 
  "audio"."audio_set_position", 
  "audio"."genres", 
  "audio"."duration", 
  "audio"."alt_files", 
  "audioset"."id", 
  "audioset"."created_on", 
  "audioset"."updated_on", 
  "audioset"."foreign_identifier", 
  "audioset"."title", 
  "audioset"."foreign_landing_url", 
  "audioset"."creator", 
  "audioset"."creator_url", 
  "audioset"."thumbnail", 
  "audioset"."provider", 
  "audioset"."url", 
  "audioset"."filesize", 
  "audioset"."filetype", 
  "api_matureaudio"."updated_on", 
  "api_matureaudio"."created_on", 
  "api_matureaudio"."media_obj_id" 
FROM 
  "audio" 
  LEFT OUTER JOIN "audioset" ON (
    "audio"."audio_set_foreign_identifier" = "audioset"."foreign_identifier" 
    AND "audio"."provider" = "audioset"."provider"
  ) 
  LEFT OUTER JOIN "api_matureaudio" ON (
    "audio"."identifier" = "api_matureaudio"."media_obj_id"
  ) 
WHERE 
  "audio"."identifier" = 'f97384f6-a5ba-4668-881f-f616904651f0' :: uuid 
LIMIT 
  21;
```

## Testing Instructions

<img width="2672" alt="Screenshot 2022-12-12 at 12 47 41 PM" src="https://user-images.githubusercontent.com/16580576/207001292-67104079-4902-45ce-bb79-f1d643d49887.png">

0. Enable logging for Django DB queries as explained [here](https://stackoverflow.com/a/20161527). 
1. You can also define a formatter with some text like "[db]" for filtering the output using `grep`.
2. Start logging the output: `just dc logs -f web | grep -E '\[db]'`.
3. Visit the search results and single result pages of audio and image media types.
4. You should see a single query for each page view in the logs.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
